### ISSUE TITLE
Editing monthly recurring transactions that have already run causes first recurrence to be skipped.

### ISSUE THREAD

**Issue by szanzibar** [firefly-iii/firefly-iii#7788]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I have some monthly recurring transactions. If I edit the transactions to be a different date in the month, the next recurrence is skipped. I suspect this is because the transaction already ran early in the month. However, even if I switch it to a daily recurrence, the first recurrence won't happen until the first of next month.

### Debug information

Debug information generated at 2023-07-27 13:22:12 Europe/Zurich for Firefly III version **6.0.18**.

| Scope | Version |
| --- | --- |
| Firefly III | 6.0.18 |
| Firefly III API | 2.0.4 |
| Build | 537, 18-07-2023 05:11:37 UTC |
| Base Build | 12, 15-07-2023 18:56:37 CEST |
| DB version | 19 (exp. 19) |
| PHP | `8.2.8` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | Europe/Zurich |
| Browser TZ | Europe/Berlin |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Max upload | 67108864 (64 MB) |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Trusted proxies (.env) | `**` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start | 2023-07-01 00:00:00 |
| Session end | 2023-07-31 23:59:59 |
| Session first | 2023-05-17 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36 |

### Expected behaviour

If I update a recurring transaction, I would expect the date selected to be the first run date. (assuming it's in the future.) I would expect this behavior regardless of the run history of the recurring transaction.

### Steps to reproduce

This transaction was a monthly transaction that ran July 1st: 
![image](https://github.com/firefly-iii/firefly-iii/assets/8921376/368495d0-126a-4dbc-8086-5ade39463cfc)

I then decided I would rather it run at the end of the month instead of the beginning, so I changed it to this:
![image](https://github.com/firefly-iii/firefly-iii/assets/8921376/63957bce-dfc8-4b8d-803a-8bf4a44c80ff)

The calendar shows what I would expect:
![image](https://github.com/firefly-iii/firefly-iii/assets/8921376/b4beb982-4c8a-4375-b49b-390c96b82fc2)

But the next date is in August:
![image](https://github.com/firefly-iii/firefly-iii/assets/8921376/35ca050d-326c-44e7-89d6-8a1e314ce95c)

This actually happened last month too after making tweaks. My transactions didn't run but I couldn't figure out why, so I just ran them all manually.

If I switch it to a daily task, it looks correct in the calendar, but the Next due is August 1st, even though that's 4 days away.

If I create a brand new task, I can't reproduce the issue. I assume it has to run once first or something?

**Comment by JC5** [comment]
Thanks, I'll check it out!

**Comment by JC5** [comment]
I can't seem to replicate this. 

> If I update a recurring transaction, I would expect the date selected to be the first run date. (assuming it's in the future.) I would expect this behavior regardless of the run history of the recurring transaction.

The code does this. It ignores past runs.

I set up a monthly recurring transaction on the 1st as you do. Today is June 29.

- If I change the monthly occurrence to 29 July or lower, the date jumps to August because it will not fire today or earlier.
- Changing the date to 30 July works as expected, first recurrence is July 30th (tomorrow).
- Changing the date to 31 also works, and corrects for September to september 30th.

Then some variants:

- Changing the recurrence to "daily" makes no difference. The behavior is the same, except it offers a daily repeat.
- A recurrence with an existing run on July 1st behaves the same.
- A recurrence with an existing run on July 15th behaves the same.
- A recurrence with an existing run on July 30 behaves the same, except the "create right now" button is hidden.

Your logs may show some info on what it calculates.

**Comment by szanzibar** [comment]
I really appreciate you looking into this.

Yeah, I actually saw the same behavior as you when I made a new recurring transaction. Unfortunately I'm only seeing this on recurring transactions that already ran earlier this month. This makes it very hard to reproduce and troubleshoot. If you set one of your new recurring rules to August 1st, then later this week after it runs, try to change to later in August, is the first recurrence skipped?

I will do the same to see if I can reproduce it again next month. I can't test today, since tomorrow is a new month, which was working before.

When I test again later this week I'll set my log level to debug and see if it yields anything helpful.

Side note, I'm a software engineer by trade, so I thought I might try to do some of my own debugging. Problem is I have 0 experience with PHP. To run locally I assume I set the env variables, start the backend, (I'll google to figure out how), and start the front end. Anything else I should know? (I didn't see any developer guidelines in the docs or repo besides https://github.com/firefly-iii/firefly-iii#contributing)
